kedro-org/kedro
# Add globals feature for `OmegaConfigLoader` using a globals resolver

## Description

Resolve #2794

## Development notes

* Added `globals` patterns to `config_patterns`
     * Can be changed/added in your project's `settings.py` through `CONFIG_LOADER_ARGS` NOTE: this is already present in `OmegaConfigLoader` and this step is only needed if users want a different pattern
    ```python
    CONFIG_LOADER_ARGS = {
        "config_patterns" = {
            "globals": ["globals*", <etc>]
         }
     }
    ```
     ❓ [QUESTION]  In the attached issue, I've mentioned passing globals patterns as separate argument "globals_pattern" and not as a part of "config_patterns". Does this make more sense? 

~~Try loading globals as a `dict` and assign to `self._globals` when `OmegaConfigLoader` is initialised.~~
  
SECTION BELOW NOT RELEVANT ANYMORE
>   * This needed me to do a some changes to the `__getitem__()` function since I couldn't use `self["globals"]` inside the `__init__` function without causing weird issues. So I separated the actual loading logic into `_get_config(key)` which is called through `__getitem__(key)` 
>     * Did some refactoring of `__getitem__()` and `load_and_merge_dir()` functions while I was at it. ❓ [QUESTION] Should I keep this or can also revert this? (I was trying a different approach which is why I separated the `_get_filtered_conf_paths()` and `_load_config_files()` initially but it is not strictly required for this implementation)

* In the `__init__` constructor -> `self._register_globals_resolver()` is called which registers a "globals" custom resolver to `OmegaConf` which calls the `_get_globals_value()` to resolve the value of the variable from `config_loader["globals"]` this happens when the config is loaded and merged by `load_and_merge_dir_config()`


 ### To Test this PR
* `settings.py` - Only if you want to change the default patterns
```python
from kedro.config import OmegaConfigLoader
CONFIG_LOADER_CLASS = OmegaConfigLoader
CONFIG_LOADER_ARGS = {
    "config_patterns": {
        "globals" : {"globals*"},
    },
}
```
* `conf/base/parameters.yml` (or any other conf file)
```yaml
my_param : "${globals:my_global_var}"
```
* `conf/base/globals.yml` 
```yaml
my_global_var: 34
```

❓ [QUESTION] Need feedback on the overwriting behaviour of globals - currently this is the same as `TemplatedConfigLoader`
* if I have globals in `local` they will override any globals in `base`
* If I have a global in `local` that doesn’t exist in the `globals.yml` in base then it should still be used in any `catalog.yml` or `parameters.yml` in `base`
* If I run `kedro run --env=base`  none of the `globals.yml` in `local` should be used because the `base` environment has been explicitly selected

### [TODO]
Waiting for feedback on the general approach before doing the following tasks -
- [x] More tests

## Checklist
